tests/e2e/state_rapid_test.go

```diff
@@ -0,0 +1,242 @@
+package main
+
+import (
+	"testing"
+
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	"pgregory.net/rapid"
+)
+
+// This file contains tests for serialization/deserialization of state.
+// The tests are written using the rapid testing library, which allows us to
+// generate arbitrary state structs and test that they can be serialized and
+// deserialized without error.
+// The generators for the various parts of the state are defined in this file, and
+// essentially tell rapid how to build the state.
+
+func TestChainStateMarshalling(t *testing.T) {
+	rapid.Check(t, func(t *rapid.T) {
+		chainState := GetChainStateGen().Draw(t, "ChainState")
+		err := MarshalAndUnmarshalChainState(chainState)
+		if err != nil {
+			t.Fatalf("error marshalling and unmarshalling chain state: %v", err)
+		}
+	})
+}
+
+// Below this are utility functions for Rapid that define generators for the various structs that can appear in testing.
+// These are used in the rapid tests and generate arbitrary test traces for fuzzing.
+// These traces will not in general be useful to execute as e2e tests, since they are filled with essentially completely random values.
+func GetStateGen() *rapid.Generator[State] {
+	return rapid.Custom(func(t *rapid.T) State {
+		return rapid.MapOf(GetChainIDGen(), GetChainStateGen()).Draw(t, "State")
+	})
+}
+
+func GetChainStateGen() *rapid.Generator[ChainState] {
+	return rapid.Custom(
+		func(t *rapid.T) ChainState {
+			valBalances := GetValBalancesGen().Draw(t, "ValBalances")
+			proposals := GetProposalsGen().Draw(t, "Proposals")
+			valPowers := GetValPowersGen().Draw(t, "ValPowers")
+			representativePowers := GetRepresentativePowersGen().Draw(t, "RepresentativePowers")
+			params := GetParamsGen().Draw(t, "Params")
+			rewards := GetRewardsGen().Draw(t, "Rewards")
+			consumerChains := GetConsumerChainsGen().Draw(t, "ConsumerChains")
+			assignedKeys := GetAssignedKeysGen().Draw(t, "AssignedKeys")
+			providerKeys := GetProviderKeysGen().Draw(t, "ProviderKeys")
+			consumerChainQueueSizes := GetConsumerChainQueueSizesGen().Draw(t, "ConsumerChainQueueSizes")
+			globalSlashQueueSize := rapid.Uint().Draw(t, "GlobalSlashQueueSize")
+
+			return ChainState{
+				ValBalances:             &valBalances,
+				Proposals:               &proposals,
+				ValPowers:               &valPowers,
+				RepresentativePowers:    &representativePowers,
+				Params:                  &params,
+				Rewards:                 &rewards,
+				ConsumerChains:          &consumerChains,
+				AssignedKeys:            &assignedKeys,
+				ProviderKeys:            &providerKeys,
+				ConsumerChainQueueSizes: &consumerChainQueueSizes,
+				GlobalSlashQueueSize:    &globalSlashQueueSize,
```

Comment:
Do we need
```
RegisteredConsumerRewardDenoms *[]string
ClientsFrozenHeights           *map[string]clienttypes.Height
```
as well? I see `RegisteredConsumerRewardDenoms` exists in `json_utils.go`.